### PR TITLE
AB#107338: Wordlist post

### PR DESCRIPTION
**Pull request by MoizIqbal97**
## Overview
Post Endpoint which Gets or Creates a wordlist

### CODE REVIEW

File: app/Decsys/Repositories/LiteDb/LiteDbWordlistRepository.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+using LiteDB;
+
+
+namespace Decsys.Repositories.LiteDb;
+
+public class LiteDbWordlistRepository : IWordlistRepository
+{
+    public UserWordlist List(string ownerId) 
+    {
+        throw new NotImplementedException();
```
Comment (AndyRae) [inline app/Decsys/Repositories/LiteDb/LiteDbWordlistRepository.cs:15]: Worth adding a comment as to why these aren't implemented. At the top of the class if it's going to be every method. 
I'm guessing it's as word list management won't be in the workshop mode? 
Otherwise if it's just implemented in a future PR take this comment as it being written down somewhere and no need to add TODOs

File: app/Decsys/Data/Entities/UserWordlist.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+using Decsys.Models.Wordlist;
+using MongoDB.Bson;
+
+namespace Decsys.Data.Entities;
+
+public class UserWordlist
+{
+    public ObjectId Id { get; set; }
```
Comment (AndyRae) [inline app/Decsys/Data/Entities/UserWordlist.cs:8]: Just a question on `ObjectId` - as it's not used for Ids elsewhere in Decsys, probably due to the Mongo/Litedb implementation? Or is it okay to use it like this?

But if this entity is exclusively for Mongo it should then be in that entity folder also.

File: app/Decsys/Controllers/WordlistsController.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        public WordlistsController(WordlistService service, IMapper mapper)
+        {
+            _service = service;
+            _mapper = mapper;
+        }
+
+
+        [HttpPost]
+        [Authorize(Policy = nameof(AuthPolicies.IsSurveyAdmin))]
+        [SwaggerOperation("Get or create a wordlist for the current user")]
```
Comment (dcl10) [inline app/Decsys/Controllers/WordlistsController.cs:28]: I don't think a POST method should 'get' a wordlist. That's what GET requests are for. Instead of `Ok` use 'CreatedAtAction` (https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.controllerbase.createdataction?view=aspnetcore-6.0)

**Review by dcl10 — CHANGES_REQUESTED** [review]

File: app/Decsys/Data/Entities/UserWordlist.cs (hunk 2 above)
Comment (MoizIqbal97) [reply app/Decsys/Data/Entities/UserWordlist.cs:8]: Using int Id started the Id at 0, and then you have to manually update them, where as mongoDb objectId makes more sense as it creates a unique 12 byte identifier.

File: app/Decsys/Controllers/WordlistsController.cs (hunk 3 above)
Comment (beforan) [reply app/Decsys/Controllers/WordlistsController.cs:28]: Yeah, this isn't the most restful interface, but it's temporarily this way for a reason.

Ordinarily, yes, you'd have a GET request to 'get' a wordlist by its id.

And you'd have a POST request perhaps to create a new wordlist, and it would be highly appropriate to use `CreatedAtAction()` to return `HTTP 201 Created` with a link to the resource where you can GET it.

Sadly, none of that really applies here.

We are temporarily rolling the two bits of functionality into one endpoint, because the behaviour we want is:
  - Create if not exists
  - Always return it, whether it was created now or previously.

Because of this, it is a "get" action really, but not of a restful resource (we don't include the id because we don't know if there even is one).

Because it has potential non idempotent side effects, POST is the safest verb, despite the non restful behaviour.

Because we don't always create, we currently always return `200 OK`. I'm open to conditionally returning `201 Created` when that's true, but there's no material gain from a consumer's perspective; they don't care about differentiating between it existing prior to this point or not; they just care about getting it.

Basically you're totally correct, but I'm not proposing we do anything about it :D

Later iterations of this functionality will support multiple wordlists with unique id's and then everything will behave restfully.

Comment (dcl10) [reply app/Decsys/Controllers/WordlistsController.cs:28]: Thanks for the explanation :)

File: app/Decsys/Data/Entities/UserWordlist.cs (hunk 2 above)
Comment (beforan) [reply app/Decsys/Data/Entities/UserWordlist.cs:8]: Oooh. First of all great q and second of all great catch.

So to answer the q: the type of id's used in DECSYS is (mostly) fairly carefully thought out, but it's not always clear due to differences in how mongo/litedb behave compared with SQL / EF in particular.

- Any document in a collection even with no apparent `Id` property in the Entity gets automatically given an `ObjectId` `_id` field by Mongo / LiteDb
- Decsys does actually use `ObjectId` (not very noticeably) for the `DecsysUser` Entity, which ultimately inherits from an `IdentityUser` with an `ObjectId` `Id`
- Neither Mongo nor LiteDb autoincrement integer id's by default, and for some Entities (e.g. Pages, Components, and now Wordlists) we want autogenerated unique ids.
    - `ObjectId` would actually be a great fit for this, but at the time I was writing LiteDb code, not Mongo, and I didn't know it had its own `ObjectId` so I used .NET's GUID and abused property initialisers.
    - a side effect of this is that the types using the GUID Id's are still portable between Mongo and LiteDb though...
- There's value in using integers sometimes. Survey and SurveyInstance use integer id's deliberately *despite* the annoyance of Mongo / LiteDb not providing built in autoincrementing, because we want them to be easily encodable into friendly access URLs (where we encode the integer id of each into a base35 string and concatenate them with `z`)

With all that in mind, we decided for Wordlists that ObjectId was a good fit in part as we are really only implementing this for Mongo, and also I had forgotten how Pages / Components were done 😅

The upshot of all this is:

- If I did it all again, or if it were Mongo only, I'd use ObjectId's instead of GUIDs
- I guess now we need to decide whether to consistently use GUID, or use native ObjectId this time.

If we use GUID:
- update the type
- ensure it initialises on entity construction

If we use ObjectId:
- good spot, yes that makes it a Mongo specific entity, so it needs to move

**Review by dcl10 — DISMISSED** [review]

File: app/Decsys/Data/Entities/UserWordlist.cs (hunk 2 above)
Comment (beforan) [reply app/Decsys/Data/Entities/UserWordlist.cs:8]: > But my vote would be for using ObjectId, given that's the ideal state.

Think I agree.

> If we were to add this functionality to workshop mode (excising Litedb feels more likely though), it looks straightforward to handle it gracefully.

Absolutely, we have an existing pattern for that with other db implementation specific entities today. :)

**Review by AndyRae — APPROVED** [review]

**Review by beforan — APPROVED** [review]
